Add unified way of initializing classes via string and configuring them.

FelixEngl opened the pull request:
Hello @jnioche,

I extracted the code for initializing classes from string and configuring them from my fork https://github.com/FelixEngl/storm-crawler/tree/local_version. (This branch contains all fixes/changes that i made. I'll try to extract PR after PR until both are on a same level.)

I don't know how far you got with #937, but I this one has a more reasonable size with the trade-off that it introduces some new warnings due to missing `@Contract`, `@NotNull` and `@Nullable` annotations in various sub-classes and sub-interfaces. These warnings will be fixed by either #937 or a PR in the future.

Best Regards

Felix


Signed-off-by: Felix Engl <felix.engl@hotmail.com>


Developer Certificate of Origin
===============================

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

jnioche commented:
hi @FelixEngl did you see my comments in #937? Some of the changes you are proposing here would meet the same objections e.g. use of org.jetbrains annotations.

FelixEngl commented:
> hi @FelixEngl did you see my comments in #937? Some of the changes you are proposing here would meet the same objections e.g. use of org.jetbrains annotations.

No, where are these comments? I didn't see them in that PR-Conversation and I didn't get any kind of notification. Or are there any other tabs that I have to check? 
That is my first time working with other people soely with GitHub, usually I'm working alone or in little groups of 5-7 people in the same room. So using GitHub for Code-Reviews is quite new to me. 
(Maybe I messed up with my notification-settings? I'll check them up later.)

jnioche commented:
> No, where are these comments? I didn't see them in that PR-Conversation and I didn't get any kind of notification. Or are there any other tabs that I have to check for notifications like that?

They were pending, you should be able to see them now

jnioche commented:
No rest for the wicked ;-)

Could you please explain what you are trying to achieve in the description of this PR and why you think it is necessary?

FelixEngl commented:
> No rest for the wicked ;-)
> 
> Could you please explain what you are trying to achieve in the description of this PR and why you think it is necessary?

Yes, the next one will be a little bit better with the description. I didn't know it better until recently. 
But I learned a lot from our discussions in #937. Thanks again for that opportunity and your patience! :)

Basically: 
A lot of classes of SC use some kind of initialisation where they initialize a class by string name. But all of them just copy-paste the very same code or a very similar one. Some of the copied codes are fautly or at least error prone without any useful information for the user/developer when they fail like "failed to initialize".

This PR proposes a unified toolkit for initializing classes from qualified names. The core of this PR is the  `InitialisationUtil`-class.

Proposing a universal way has multiple advantages:
- No more copy paste code. 
- A clean and direct approach that works for everyone.
- If the way how a class is initialized by string changes, we can simply adapt the change at a single place and not everywhere.
- The `InitialisationUtil` is even useable from developers using SC, therefore writing clean code/extensions by utilizing this for initializing from string names leads to a more secure and clean code in the future.
- Due to the importance of that feature I decided to create fine grained exceptions with easy to understand messages, making the development/debugging for and with SC much more easier.
- The new `InitialisationUtil` even allows for additional checks to assert, that the initialized class does not just implement the said superclass but also some other interface.

FelixEngl commented:
@jnioche Ah, I forgot something:
The other notable change is at core/src/main/java/com/digitalpebble/stormcrawler/protocol/ProtocolFactory.java where I introduce double checket locking for a more secure access of the singleton in a multithreaded context.
Otherwise there could be cases where multiple instances of the singleton are created.

rzo1 commented on core/src/main/java/com/digitalpebble/stormcrawler/parse/JSoupFilters.java:
```diff
@@ -75,11 +75,7 @@ public static JSoupFilters fromConf(Map<String, Object> stormConf) {
-     *
-     * @throws IOException
-     */
+    /** loads the filters from a JSON configuration file */
```
This might break end user code (removal of a checked exception), so we need to be clear in the changelog.

FelixEngl replied:
Why should this break end user code? 
It still throws that `IOException` but i removed the empty comment.

rzo1 commented on core/src/main/java/com/digitalpebble/stormcrawler/persistence/Scheduler.java:
```diff
@@ -43,20 +43,10 @@ public static Scheduler getInstance(Map<String, Object> stormConf) {
+            scheduler = InitialisationUtil.initializeFromQualifiedName(className, Scheduler.class);
         } catch (Exception e) {
-            throw new RuntimeException("Can't instanciate " + className);
+            throw new RuntimeException("Can't instanciate " + className, e);
```
Good catch!

rzo1 commented on core/src/main/java/com/digitalpebble/stormcrawler/persistence/urlbuffer/URLBuffer.java:
```diff
@@ -85,29 +89,4 @@ default void acked(String url) {
     default void configure(Map<String, Object> stormConf) {}
-
-    /** Returns a URLBuffer instance based on the configuration * */
-    public static URLBuffer getInstance(Map stormConf) {
```
This might break end user code (removal of a public method), so we need to be clear in the changelog.

rzo1 commented on core/src/main/java/com/digitalpebble/stormcrawler/util/ConfigurableUtil.java:
```diff
@@ -0,0 +1,128 @@
+package com.digitalpebble.stormcrawler.util;
```
Misses a license header

jnioche replied:
yes every new class needs a license header

rzo1 commented on core/src/main/java/com/digitalpebble/stormcrawler/util/exceptions/initialisation/ClassForInitialisationNotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+package com.digitalpebble.stormcrawler.util.exceptions.initialisation;
```
License? wdyt @jnioche

jnioche replied:
yes every new class needs a license header

FelixEngl replied:
@jnioche What about Tests?

jnioche replied:
yes, test classes as well please.

rzo1 commented on core/src/main/java/com/digitalpebble/stormcrawler/util/exceptions/initialisation/ClassNotAssignableException.java:
```diff
@@ -0,0 +1,15 @@
+package com.digitalpebble.stormcrawler.util.exceptions.initialisation;
```
License? wdyt @jnioche

rzo1 commented on core/src/main/java/com/digitalpebble/stormcrawler/util/exceptions/initialisation/InitialisationException.java:
```diff
@@ -0,0 +1,19 @@
+package com.digitalpebble.stormcrawler.util.exceptions.initialisation;
```
License? wdyt @jnioche

rzo1 commented on core/src/main/java/com/digitalpebble/stormcrawler/util/exceptions/initialisation/NoUsableConstructorFoundException.java:
```diff
@@ -0,0 +1,11 @@
+package com.digitalpebble.stormcrawler.util.exceptions.initialisation;
```
License? wdyt @jnioche

rzo1 commented on core/src/main/java/com/digitalpebble/stormcrawler/util/exceptions/initialisation/NotAssignableException.java:
```diff
@@ -0,0 +1,39 @@
+package com.digitalpebble.stormcrawler.util.exceptions.initialisation;
```
License? wdyt @jnioche

rzo1 commented on core/src/main/java/com/digitalpebble/stormcrawler/util/exceptions/initialisation/NotInitializeableException.java:
```diff
@@ -0,0 +1,7 @@
+package com.digitalpebble.stormcrawler.util.exceptions.initialisation;
```
License? wdyt @jnioche

rzo1 commented on core/src/main/java/com/digitalpebble/stormcrawler/util/exceptions/initialisation/QualifiedClassNameBlankException.java:
```diff
@@ -0,0 +1,7 @@
+package com.digitalpebble.stormcrawler.util.exceptions.initialisation;
```
License? wdyt @jnioche

rzo1 commented on core/src/main/java/com/digitalpebble/stormcrawler/util/exceptions/initialisation/SuperclassIsPrimitiveException.java:
```diff
@@ -0,0 +1,14 @@
+package com.digitalpebble.stormcrawler.util.exceptions.initialisation;
```
License? wdyt @jnioche

rzo1 commented on core/src/main/java/com/digitalpebble/stormcrawler/util/exceptions/initialisation/SuperclassNotAssignableException.java:
```diff
@@ -0,0 +1,18 @@
+package com.digitalpebble.stormcrawler.util.exceptions.initialisation;
```
License? wdyt @jnioche

rzo1 commented on core/src/test/java/com/digitalpebble/stormcrawler/helper/initialisation/ClassInheritingFomAbstractAndInterface.java:
```diff
@@ -0,0 +1,7 @@
+package com.digitalpebble.stormcrawler.helper.initialisation;
```
License? wdyt @jnioche

rzo1 commented on core/src/test/java/com/digitalpebble/stormcrawler/helper/initialisation/ClassInheritingFromAbstractClassOnly.java:
```diff
@@ -0,0 +1,5 @@
+package com.digitalpebble.stormcrawler.helper.initialisation;
```
License? wdyt @jnioche

rzo1 commented on core/src/test/java/com/digitalpebble/stormcrawler/helper/initialisation/ClassInheritingFromAbstractClassOnly.java:
```diff
@@ -0,0 +1,5 @@
+package com.digitalpebble.stormcrawler.helper.initialisation;
+
```
License? wdyt @jnioche

rzo1 commented on core/src/test/java/com/digitalpebble/stormcrawler/helper/initialisation/ClassInheritingFromOpenClass.java:
```diff
@@ -0,0 +1,5 @@
+package com.digitalpebble.stormcrawler.helper.initialisation;
```
License? wdyt @jnioche

rzo1 commented on core/src/test/java/com/digitalpebble/stormcrawler/helper/initialisation/ClassWithoutValidConstructor.java:
```diff
@@ -0,0 +1,9 @@
+package com.digitalpebble.stormcrawler.helper.initialisation;
```
License? wdyt @jnioche

rzo1 commented on core/src/test/java/com/digitalpebble/stormcrawler/helper/initialisation/ClassWithoutValidConstructor.java:
```diff
@@ -0,0 +1,9 @@
+
+public class ClassWithoutValidConstructor extends AbstractClass {
+    public ClassWithoutValidConstructor(int i) {
+        System.out.println(i);
```
What is `i` and why sysout? Might be difficult to check against sysout in unit tests ;)

rzo1 commented on core/src/test/java/com/digitalpebble/stormcrawler/helper/initialisation/FinalClassToInitialize.java:
```diff
@@ -0,0 +1,3 @@
+package com.digitalpebble.stormcrawler.helper.initialisation;
```
License? wdyt @jnioche

rzo1 commented on core/src/test/java/com/digitalpebble/stormcrawler/helper/initialisation/SimpleOpenClass.java:
```diff
@@ -0,0 +1,3 @@
+package com.digitalpebble.stormcrawler.helper.initialisation;
```
License? wdyt @jnioche

rzo1 commented on core/src/test/java/com/digitalpebble/stormcrawler/helper/initialisation/base/AbstractClass.java:
```diff
@@ -0,0 +1,21 @@
+package com.digitalpebble.stormcrawler.helper.initialisation.base;
```
License? wdyt @jnioche

rzo1 commented on core/src/test/java/com/digitalpebble/stormcrawler/helper/initialisation/base/ITestInterface.java:
```diff
@@ -0,0 +1,3 @@
+package com.digitalpebble.stormcrawler.helper.initialisation.base;
```
License? wdyt @jnioche

rzo1 commented on core/src/test/java/com/digitalpebble/stormcrawler/helper/initialisation/base/OpenClassWithAbstractClassAndInterface.java:
```diff
@@ -0,0 +1,4 @@
+package com.digitalpebble.stormcrawler.helper.initialisation.base;
```
License? wdyt @jnioche

rzo1 commented on core/src/test/java/com/digitalpebble/stormcrawler/util/InitialisationUtilTest.java:
```diff
@@ -0,0 +1,168 @@
+package com.digitalpebble.stormcrawler.util;
```
License? wdyt @jnioche

jnioche commented:
Guys, apologies in advance if I come across as being a bit blunt.
I am totally in favour of improving aspect of the code but this will not come at the cost of adding I-don't-know-how-many classes for various exceptions + tons if utility classes and so on, for what is not a new functionality that users will benefit from or a major improvement.
One of the things that I am pleased to have achieved with SC is (I think) a good trade-off between code simplicity and richness of features. I intend to keep it that way.
Can this be done in a way that doesn't involve the addition of so many new classes?
Thanks

FelixEngl commented:
> Guys, apologies in advance if I come across as being a bit blunt. I am totally in favour of improving aspect of the code but this will not come at the cost of adding I-don't-know-how-many classes for various exceptions + tons if utility classes and so on, for what is not a new functionality that users will benefit from or a major improvement. One of the things that I am pleased to have achieved with SC is (I think) a good trade-off between code simplicity and richness of features. I intend to keep it that way. Can this be done in a way that doesn't involve the addition of so many new classes? Thanks

Don't worry about being blunt, I prefer it when someone speaks his/her mind directly instead of dancing around the problem or finding some straw arguments. :D

I could remove the exception classes and only throw runtime exceptions. That would reduce the number of classes.
Would that be enough or are there other classes that are too much too?

rzo1 commented:
I think, that it looks like a lot of classes. Basically, I see 2 (new) utility classes + a lot of test-related classes + (runtime) exception sugar. Guess, that the runtime exception can be dropped.

FelixEngl commented:
@jnioche I removed the Exceptions. We could also remove the `URLBufferUtil` if you want.
Then we would basically only introduce one new class in the code.

But we need the other new classes for testing the initialisation feature, so we can't remove the new classes created in the test source.

FelixEngl commented:
> @jnioche I removed the Exceptions. We could also remove the `URLBufferUtil` if you want. Then we would basically only introduce one new class in the code.
> 
> But we need the other new classes for testing the initialisation feature, so we can't remove the new classes created in the test source.

Removed the `URLBufferUtil` too, now we have only one new class.

jnioche commented on core/src/main/java/com/digitalpebble/stormcrawler/protocol/DelegatorProtocol.java:
```diff
@@ -156,19 +144,27 @@ boolean isMatch(final Metadata metadata) {
     private final LinkedList<FilteredProtocol> protocols = new LinkedList<>();
```
@FelixEngl could we replace the LinkedList with an ArrayList or ArrayDeque as suggested by Lift?

jnioche commented:
> > @jnioche I removed the Exceptions. We could also remove the `URLBufferUtil` if you want. Then we would basically only introduce one new class in the code.
> > But we need the other new classes for testing the initialisation feature, so we can't remove the new classes created in the test source.
> 
> Removed the `URLBufferUtil` too, now we have only two new classes. `InitialisationUtil` and `ConfigurableUtil`.

Great! That's way better

_InitialisationUtil_ has no equivalent in the current code, so we definitely need it.  _ConfigurableUtil_ is a bit different. I see that you removed the utility code from the interface and kept the latter as a strict interface. I can understand why but to be honest I quite liked having the utility code within the interface class - all in one place. Probably not orthodox, I agree but I thought it was quite elegant in a way. How open would you be to keeping it that way? This is just an extra class, I can also be swayed.

And @rzo1 is right, we need license headers on every new file.

Thanks for your open-mindedness BTW.

jnioche commented on core/src/main/java/com/digitalpebble/stormcrawler/protocol/ProtocolFactory.java:
```diff
@@ -26,14 +27,31 @@
     public static ProtocolFactory getInstance(Config conf) {
 
-        if (single_instance != null) return single_instance;
+        // https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_Java
```
such a great bugfix! thanks for that

FelixEngl commented:
> > > @jnioche I removed the Exceptions. We could also remove the `URLBufferUtil` if you want. Then we would basically only introduce one new class in the code.
> > > But we need the other new classes for testing the initialisation feature, so we can't remove the new classes created in the test source.
> > 
> > 
> > Removed the `URLBufferUtil` too, now we have only two new classes. `InitialisationUtil` and `ConfigurableUtil`.
> 
> Great! That's way better
> 
> _InitialisationUtil_ has no equivalent in the current code, so we definitely need it. _ConfigurableUtil_ is a bit different. I see that you removed the utility code from the interface and kept the latter as a strict interface. I can understand why but to be honest I quite liked having the utility code within the interface class - all in one place. Probably not orthodox, I agree but I thought it was quite elegant in a way. How open would you be to keeping it that way? This is just an extra class, I can also be swayed.
> 
> And @rzo1 is right, we need license headers on every new file.
> 
> Thanks for your open-mindedness BTW.

I didn't have a problem with the static function (tbh I like putting creation calls for interfaces there too) but i had a problem with the static `LOG` reference in the interface. That was used by one of subclasses, and that is quite confusing when logging and you only see the interface-reference.

What about hiding the ConfigUtil in package scope and keeping the original creation method of the interface?
That way we can assure that the Logger for the `configure` is only called by the `Configure` and nowhere else.

FelixEngl commented:
@jnioche I implemented the hidden-util-class approach, I think this should make both of us happy.
Furthermore I added to all files the license and replaced the LinkedList.
And you are right, it feels more natural when you use a static function in the interface for that instead of an until-class. :)

jnioche reviewed (changes requested):
Looks good, just one minor point about returning an immutable List
